ARCHITECTURE.md

```diff
@@ -0,0 +1,275 @@
+assists with finding the shortest path to the desired key.
+
+When a node is written, its *prefix hash* is computed. This done by first
+splitting the key into its components (`a`, `b`, and `c` for `/a/b/c`), and then
+hashing each component into a 32-character hash, where one character is a 2-bit
+value (0, 1, 2, or 3). The `prefix` hash for `/a/b/c` is
+
+```js
+node.path = [
+1, 2, 0, 1, 2, 0, 2, 2, 3, 0, 1, 2, 1, 3, 0, 3, 0, 0, 2, 1, 0, 2, 0, 0, 2, 0, 0, 3, 2, 1, 1, 2,
+0, 1, 2, 3, 2, 2, 2, 0, 3, 1, 1, 3, 0, 3, 1, 3, 0, 1, 0, 1, 3, 2, 0, 2, 2, 3, 2, 2, 3, 3, 2, 3,
+0, 1, 1, 0, 1, 2, 3, 2, 2, 2, 0, 0, 3, 1, 2, 1, 3, 3, 3, 3, 3, 3, 0, 3, 3, 2, 3, 2, 3, 0, 1, 0,
+4 ]
+```
+
+Each component is divided by a newline. `4` is a special value indicating the
+end of the prefix.
+
+#### Example
+
+Consider a fresh HyperDB. We write `/a/b = 24` and get back this node:
+
+```js
+{ key: '/a/b',
+  value: '24',
+  clock: [ 0 ],
+  trie: [],
+  feeds: [ [Object] ],
+  feedSeq: 0,
+  feed: 0,
+  seq: 0,
+  path: 
+   [ 1, 2, 0, 1, 2, 0, 2, 2, 3, 0, 1, 2, 1, 3, 0, 3, 0, 0, 2, 1, 0, 2, 0, 0, 2, 0, 0, 3, 2, 1, 1, 2,
+     0, 1, 2, 3, 2, 2, 2, 0, 3, 1, 1, 3, 0, 3, 1, 3, 0, 1, 0, 1, 3, 2, 0, 2, 2, 3, 2, 2, 3, 3, 2, 3,
+     4 ] } 
+```
+
+If you compare this path to the one for `/a/b/c` above, you'll see that the
+first 64 2-bit characters match. This is because `/a/b` is a prefix of `/a/b/c`.
+
+Since this is the first entry, `seq` is 0. Since this is the only known feed,
+`feed` is also 0. `feeds` is an array of entries of the form `{ key: Buffer,
+seq: Number }` that let you map the numeric value `feed` to a hypercore key and
+its sequence number head. `feeds` isn't always set: it only gets included when
+it changes compared to `node.seq - 1`, in the interest of storing less data per
+node.
+
+Now we write `/a/c = hello` and get this node:
+
+```js
+{ key: '/a/c',
+  value: 'hello',
+  clock: [ 0 ],
+  trie: [ , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , [ , , [ { feed: 0, seq: 0 } ] ] ],
+  feeds: [],
+  feedSeq: 0,
+  feed: 0,
+  seq: 1,
+  path: 
+   [ 1, 2, 0, 1, 2, 0, 2, 2, 3, 0, 1, 2, 1, 3, 0, 3, 0, 0, 2, 1, 0, 2, 0, 0, 2, 0, 0, 3, 2, 1, 1, 2,
+     0, 1, 1, 0, 1, 2, 3, 2, 2, 2, 0, 0, 3, 1, 2, 1, 3, 3, 3, 3, 3, 3, 0, 3, 3, 2, 3, 2, 3, 0, 1, 0,
+     4 ] }
+```
+
+As expected, this node has the same `feed` value as before (since we're only
+writing to one feed). Its `seq` is 1, since the last was 0. Notice that `feeds`
+isn't included, because the mapping of the numeric `feed` value to a key hasn't
+changed.
+
+Also, this and the previous node have the first 32 characters of their `path` in
+common (the prefix `/a`).
+
+Notice though that `trie` is set. It's a long but sparse array. It has 35
+entries, with the last one referencing the first node inserted (`a/b/`). Why?
+
+(Really, it's a sparse array with 64 entries (the length of the `path`). But in
+Javascript only the entries preceding the last value get printed.)
```

Comment:
You have lost me here. Why in javascript is this the case?

Reply:
Same.

Reply:
Oh the penny drops. Its because if you look at an empty array it does not matter that the trailing values are empty. It prints up until the last non empty value. That is a hard thing to explain.

Reply:
I was wrong about this!

```
> var a = new Array(5)
> a
[ , , , , ]
> a[2] = 3
> a
[ , , 3, , ]
```

I rewrote this part hopefully be more clarifying:

```
(If it wasn't stored as a sparse array, you'd actually see 64 entries (the
length of the `path`). But since the other 29 entries are also empty, hyperdb
doesn't bother allocating them.)
```

Reply:
that is also much much clearer